clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -4368,16 +4348,18 @@ class SyclKernelBodyCreator : public SyclKernelFieldHandler {
 class FreeFunctionKernelBodyCreator : public SyclKernelFieldHandler {
   SyclKernelDeclCreator &DeclCreator;
   llvm::SmallVector<Stmt *, 16> BodyStmts;
+  // Keep track of the structs we have encountered on our way to a special type.
+  // They will be needed to properly generate the __init call. Note that the
```

Comment:
How are these kept for SYCL 2020 kernels? How free function kernels are different from SYCL 2020 so we need this additional tracking?

Reply:
`SyclKernelBodyCreator` uses the `MemberExprBases` vector which serves a similar purpose. I think there is slightly less code this way for example MemberExprBases requires a bit more work to maintain properly as we enter and leave structs. Also `FreeFunctionKernelBodyCreator` does not inherit from `SyclKernelBodyCreator` so we are not wasting information. I just chose a slightly different way of doing things.